llvm/lib/CAS/LazyMappedFileRegion.cpp

```diff
@@ -242,14 +242,67 @@ Expected<LazyMappedFileRegion> LazyMappedFileRegion::create(
 
   sys::fs::file_t File = sys::fs::convertFDToNativeFile(FD);
 
+  struct FileLockRAII {
+    LazyMappedFileRegion &LMFR;
+    bool IsLocked = false;
+
+    enum LockKind { Shared, Exclusive };
+
+    FileLockRAII(LazyMappedFileRegion &LMFR) : LMFR(LMFR) {}
+    ~FileLockRAII() {
+      if (Error E = unlock())
+        consumeError(std::move(E));
+    }
+
+    Error lock(LockKind LK) {
+      if (Error E = unlock())
+        return E;
+      if (std::error_code EC = sys::fs::lockFile(*LMFR.FD, LK == Exclusive))
+        return createFileError(LMFR.Path, EC);
+      IsLocked = true;
+      return Error::success();
+    }
+
+    Error unlock() {
+      if (IsLocked) {
+        IsLocked = false;
+        if (std::error_code EC = sys::fs::unlockFile(*LMFR.FD))
+          return createFileError(LMFR.Path, EC);
+      }
+      return Error::success();
+    }
+
+  } FileLock(LMFR);
+
+  // Use shared/reader locking in case another process is in the process of
+  // initializing the file.
+  if (Error E = FileLock.lock(FileLockRAII::Shared))
+    return std::move(E);
+
   sys::fs::file_status Status;
   if (std::error_code EC = sys::fs::status(File, Status))
     return errorCodeToError(EC);
-  if (Status.getSize() > 0)
+
+  if (Status.getSize() == 0) {
+    // Lock the file exclusively so only one process will do the initialization.
+    if (Error E = FileLock.lock(FileLockRAII::Exclusive))
+      return std::move(E);
+    if (std::error_code EC = sys::fs::status(File, Status))
+      return errorCodeToError(EC);
+  }
+
+  // At this point either the file is still empty (this process won the race to
+  // do the initialization) or we have the size for the completely initialized
+  // file.
+
+  if (Status.getSize() > 0) {
     // The file was already constructed.
+    if (Error E = FileLock.unlock())
```

Comment:
This might be an edge case that but you might want to hold the lock till after `mapped_file_region` is created (where RAII object naturally destruct) because the other process might be in the process of extending the file so you might get a half created content for next page and the `CacheSize` might not be consistent (?). It might totally be fine if you just reading content, but it is probably better just to do a full synchronization so we don't need to worry about those cases.

Reply:
Good point, done.👍